fix: add documentation for storage deposit and update gas fee simulation instructions (#4517)
gnolang/gno: docs/resources/storage-deposit.md

> 
> ## Overview of storage deposit

Reviewer: ```suggestion
# Storage Deposits
```

In Gno.land, storage is a paid resource. To persist data in a realm (such as
setting variables or storing objects), users must lock GNOT tokens as a storage
deposit. This ensures efficient, accountable use of on-chain storage.

Storage costs are settled per message, and tokens are locked or refunded
depending on the net change in data usage.

> ## How it works

Reviewer: No need for this
```suggestion
```

### What is a Storage Deposit?

A storage deposit is an amount of GNOT locked to pay for the storage space your
data occupies on-chain. The system calculates and deducts this amount after each
message (e.g., MsgCall, MsgRun, AddPkg).

Storing data → GNOT locked
Deleting data → GNOT refunded

### Purpose

> - Pay for persistent storage: Store objects like structs or strings in realms.
> 
> - Encourage cleanup: Reclaim deposits by deleting unneeded data.
> 
> - Flexibility: Realm developers can design their own cleanup or reward logic.

Reviewer: ```suggestion
- Pay for persistent storage: Store objects like structs or strings in realms.
- Encourage cleanup: Reclaim deposits by deleting unneeded data.
- Flexibility: Realm developers can design their own cleanup or reward logic.
```

### Storage Settlement Flow

1 Start with a message call (e.g. AddPkg)

2 Specify optional -max-deposit to limit the GNOT to be locked for storage.

3 Storage delta is calculated (how much it grew or shrunk).

4 System locks or refunds GNOT accordingly.

### Tracking Storage

Use this command to inspect current storage usage and deposit in a realm:

```bash

> gnokey query vm/qstorage --data gno.land/r/foo

Reviewer: Since we've added a new query path, please add a small section in `interact-with-gnokey.md` just like there is for all of the other ones 🙏

Reviewer: You can actually just use this whole section `### Tracking Storage` as the paragraph over there, and just link it here in some convenient spot

```

Sample Output:

```

> storage: 5025, deposit: 502500

Reviewer: Is it valid to assume that I can get the price of 1 byte of storage by doing the following

`deposit/storage = 502500/5025 = 100 ugnot`?

We should denominate both fields in general. Can we do that in the query result as well?

Reviewer: Add a line explaining that the price can also be calculated like this instead of getting it from the params realm

```

storage: total bytes used

deposit: total GNOT locked for that storage

### Anyone Can Free Storage

Because deposits are tracked per realm (not per user), anyone can remove data.
This can create a “cleanup reward” scenario where the person who deletes data
gains the released deposit. It also gives the realm developer flexibility to
design and manage user storage.

> ### Global Storage Price Parameter
> 
> The storage price is a global parameter governed by the GovDAO'
> 
> ```
> storagePriceDefault = "100ugnot" // cost per byte
> // e.g., 1 GNOT per 10KB (≈ 1B GNOT = 10TB)
> ```

Reviewer: Can you link to where the storage `storagePriceDefault` lives? Does it live in `r/sys/params`?

```suggestion
### Global Storage Price Parameter

The storage price is a global parameter governed by the GovDAO. 

```
storagePriceDefault = "100ugnot" // cost per byte
// e.g., 1 GNOT per 10KB (≈ 1B GNOT = 10TB)
```
```

## Storage deposits

In Gno.land, storage is a paid resource. To persist data in a realm (such as
setting variables or storing objects), users must lock GNOT tokens as a storage
deposit. This ensures efficient, accountable use of on-chain storage.

Storage costs are settled per message, and tokens are locked or refunded

> depending on the net change in data usage.

Author: That is correct. I will make it clear.

### What is a Storage Deposit?

A storage deposit is an amount of GNOT locked to pay for the storage space your
data occupies on-chain. The system calculates and deducts this amount after each
message (e.g., MsgCall, MsgRun, AddPkg).

Storing data → GNOT locked
Deleting data → GNOT refunded

### Purpose

> - Pay for persistent storage: Store objects like structs or strings in realms.

Reviewer: ```suggestion
- Paying for persistent storage: Storing objects or primitives in in realms costs GNOT.
```

> - Encourage cleanup: Reclaim deposits by deleting unneeded data.

Reviewer: ```suggestion
- Encouraging cleanup: Users can reclaim deposits by deleting unneeded data.
```

- Flexibility: Realm developers can design their own cleanup or reward logic.

> ### Storage Settlement Flow
> 
> 1 Start with a message call (e.g. AddPkg)
> 
> 2 Specify optional -max-deposit to limit the GNOT to be locked for storage.
> 
> 3 Storage delta is calculated (how much it grew or shrunk).
> 
> 4 System locks or refunds GNOT accordingly.
> 
> 
> ### Anyone Can Free Storage

Reviewer: ```suggestion
### Storage Settlement Flow

Below is an example of how the storage fee settlement flow works:

1. Start with a message call (e.g. AddPkg)
2. Specify optional `-max-deposit` to limit the GNOT that can be locked for storage.
3. The storage delta is calculated by the GnoVM (how much it grew or shrunk).
4. The system locks or refunds GNOT accordingly.

### Anyone Can Free Storage
```

Because deposits are tracked per realm (not per user), anyone can remove data.
This can create a “cleanup reward” scenario where the person who deletes data
gains the released deposit. It also gives the realm developer flexibility to

> design and manage user storage.

Reviewer: The question here is; why would any developer do this? Why make a cleanup function available to users, when the deployer is not actually paying for any fees? Also, the developer can make a cleanup function that only they are allowed to execute, and thus make money on their users. Isn't this a concern for the system?

Author: Really good questions!  

> why would any developer do this? Why make a cleanup function available to users, when the deployer is not actually paying for any fees?

The deposit is essentially a storage fee — a cost that users cannot get refunded. If a developer provides functions that allow users to reclaim their storage fees, it would be more cost-effective for users compared to other developers who don't offer such functionality. This could drive greater usage.

> Also, the developer can make a cleanup function that only they are allowed to execute, and thus make money on their users. Isn't this a concern for the system?

Compared to developers who, by default, do not provide any mechanism to release storage in their contracts, users cannot recover anything — the storage deposit becomes a sunk cost. In that case, it’s no worse than expected. However, when users compare this to developers who do offer the ability to clean up and reclaim storage, the absence of such a feature may be perceived as the developer “profiting off their users.” These developers may be penalized with less traction and lower usage. This dynamic creates peer pressure for developers to support storage refund capabilities, leading to positive outcomes for users, developers, and the chain as a whole.

These assumptions rely on rational and legitimate user and developer behavior. We’ll need to test and validate them once the feature is released.

### Global Storage Price Parameter

The storage price is a global parameter governed by the GovDAO'

> The default value is defined in `gno.land/pkg/sdk/vm/params.go`.

Reviewer: Use the github link here please

```
storagePriceDefault = "100ugnot" // cost per byte
// e.g., 1 GNOT per 10KB (≈ 1B GNOT = 10TB)
```

### Tracking Storage

You can inspect the current storage usage and deposit in a realm.
See more usage [examples](../users/interact-with-gnokey.md#vmqstorage).

### Example

The Clear() function removes the content stored in the realm gno.land/r/foo.

```bash

gnokey maketx call \
  -pkgpath gno.land/r/foo \
  -func Clear \
  -gas-fee 1000000ugnot \
  -gas-wanted 10000000 \
  -broadcast \
  -remote https://rpc.gno.land:443 \
  -chainid staging \
  YOUR_KEY_NAME
```

You will see output similar to the following in the event:

```
GAS WANTED: 10000000
GAS USED:   291498
EVENTS: [
  {
    "type": "UnlockDeposit",
    "attrs": [
      {"key": "Deposit", "value": "2700ugnot"},
      {"key": "ReleaseStorage", "value": "27 bytes"}
    ]
  }

> ]

Reviewer: I think we should remove this completely
